homeworks/week4/Hong-Nguyen-week4

```diff
@@ -0,0 +1,94 @@
+// Given a list of intervals [low, high) where low < high and len(intervals) >= 1, uniformly randomly generate a point from the union of these intervals.
+// For example, given intervals [2,5], [1,3], method generateRandomValue produces a random value in [1, 2) with 20%, [2,3) with 40%, and [3,5) with 40%.
+
+import java.util.Random;
+import java.util.*;
+
+class Main {
+  public static void main(String[] args) {
+    Interval[] intervals = {
+      new Interval(1,3),
+      new Interval(2,5),
+      new Interval(1,4),
+      new Interval(0,2)
+    };
+
+    int[] count = new int[5];
+    int iterations = 100000;
+    for (int i = 0; i < iterations; i++) {
+      double f = generateRandomValue(intervals);
+      for (int j = 1; j <= 5; j++)
+      if (f >= j-1 && f < j) {
+        count[j-1] ++;
+      }
+    }
+
+    for (int j = 1; j <= 5; j++) {
+      System.out.println("Interval [" + (j-1) + "," + j + ") with probability " + 100*(double)count[j-1]/iterations + " %");
+    }
+    
+    // Interval [0,1) with probability ~10%
+    // Interval [1,2) with probability ~30%
+    // Interval [2,3) with probability ~30%
+    // Interval [3,4) with probability ~20%
+    // Interval [4,5) with probability ~10%
+  }
+
+  static double[] lengthSums = new double[0];
+
+  /*
+    If number of intervals is N, then
+    time complexity is O(N) as we need to loop through 
+    it to build the cumulativeLengthSum
+    Space complexity is O(n)
```

Comment:
This solution is very similar to my sample solution. If you precompute lengthSums once, and use it many times later. Then the time complexity can be O(logN)

Please see a simpler solution and see the trade-off in term of space/time.
https://github.com/P2PLearning/DSA/pull/28/files